lib/src/main/kotlin/konig/kontext/KonigKontextServerInterceptor.kt

```diff
@@ -1,22 +1,43 @@
 package org.konigsoftware.konig.kontext
 
+import com.google.protobuf.Message
+import io.grpc.BindableService
 import io.grpc.Metadata
+import io.grpc.ServerBuilder
 import io.grpc.ServerCall
 import io.grpc.kotlin.CoroutineContextServerInterceptor
+import konig.kontext.GRPC_HEADER
 import kotlin.coroutines.CoroutineContext
+import kotlin.reflect.KClass
+import kotlin.reflect.full.functions
+import kotlin.reflect.jvm.javaMethod
+import kotlin.reflect.safeCast
 
-class KonigKontextServerInterceptor<ContextType>(private val konigKontext: KonigKontext<ContextType>) : CoroutineContextServerInterceptor() {
+class KonigKontextServerInterceptor<ContextType : Message>(
+    private val contextClass: KClass<ContextType>
+) : CoroutineContextServerInterceptor() {
     override fun coroutineContext(call: ServerCall<*, *>, headers: Metadata): CoroutineContext {
-        val coroutineKonigContext = CoroutineKonigKontext()
+        val konigKontextBinary = headers.get(GRPC_HEADER) ?: run {
+            "".toByteArray()
+        }
 
-        val grpcKonigKontextBinary = headers.get(konigKontext.GRPC_HEADER) ?: run {
-            coroutineKonigContext.setContext(konigKontext.default)
+        val parserFunction =
+            contextClass.functions.find { it.name == "parseFrom" && it.parameters.size == 1 && it.javaMethod?.parameterTypes?.first() == ByteArray::class.java }
+                ?: throw IllegalStateException("Could not find binary parser function for ContextType. Please ensure your KonigKontext type extends the com.google.protobuf.Message class")
 
-            return coroutineKonigContext
-        }
+        val parsedContextUntyped = parserFunction.call(konigKontextBinary) ?: throw IllegalStateException(
+            "Unable to parse ${
+                String(konigKontextBinary)
+            } as ${contextClass.simpleName}. Please ensure your KonigKontext type extends the com.google.protobuf.Message class"
```

Comment:
wouldn't we throw this if the message was just invalid? (but `KonigKontext` was extending the right thing)

(same below)